proto/v2/convert_proto.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 The OpenZipkin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package zipkin_proto3 adds support for the Zipkin protobuf definition to allow
+Go applications to consume model.SpanModel from protobuf serialized data.
+*/
+package zipkin_proto3
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	zipkinmodel "github.com/openzipkin/zipkin-go/model"
+)
+
+// ParseSpans parses zipkinmodel.SpanModel values from data serialized by Protobuf3.
+// debugWasSet is a boolean that toggles the Debug field of each Span. Its value
+// is usually retrieved from the transport headers when the "X-B3-Flags" header has a value of 1.
+func ParseSpans(protoBlob []byte, debugWasSet bool) (zss []*zipkinmodel.SpanModel, err error) {
+	var listOfSpans ListOfSpans
+	if err := proto.Unmarshal(protoBlob, &listOfSpans); err != nil {
+		return nil, err
+	}
+	for _, zps := range listOfSpans.Spans {
+		zms, err := protoSpanToModelSpan(zps, debugWasSet)
+		if err == nil && zms != nil {
+			// TODO: perhaps record or combine these errors?
+			zss = append(zss, zms)
+		}
+	}
+	return zss, nil
+}
+
+var errNilZipkinSpan = errors.New("expecting a non-nil Span")
+
+// This conversion is done so that we can reuse the logic for converting from
+//      Zipkin-Go SpanModel --> opencensusProto.Span
+// as was done during Zipkin-Go SpanModel JSON deserialization
+// instead of reimplementing zipkin_proto3 --> opencensusProto.Span.
+// This conversion shouldn't cause any loss of fidelity.
+func protoSpanToModelSpan(zps *Span, debugWasSet bool) (*zipkinmodel.SpanModel, error) {
+	if zps == nil {
+		return nil, errNilZipkinSpan
+	}
+	traceID, err := zipkinmodel.TraceIDFromHex(fmt.Sprintf("%x", zps.TraceId))
+	if err != nil {
+		return nil, fmt.Errorf("TraceID: %v", err)
```

Comment:
I think we should follow the errors messages styles we have in https://github.com/openzipkin/zipkin-go/blob/master/propagation/b3/shared.go. So I would rephrase this as `invalid traceID: %v` or something like that. What do you think?